py/desispec/workflow/schedule.py

```diff
@@ -0,0 +1,189 @@
+        for req in reqs:
+            if not req.Test():
+                return False
+        return True
+
+    def _schedule(self):
+
+        """
+        Schedule, run and assign processes for all jobs in this object        
+        """
+        # bookkeeping
+        waitlist        = [] # message handles for pending worker groups
+        worker_groups   = [] # worker assigned
+
+        # start by assigning Ngroups jobs to each of the Ngroup groups
+        nextjob=0
+        for job in range(self.Ngroups):
+            worker = nextjob
+            reqs=self._assign_job(worker,nextjob)
+            waitlist.append(reqs)
+            worker_groups.append(worker)
+            nextjob += 1
+
+        # the scheduler waits for jobs to be completed;
+        # when one is complete it assigns the next job
+        # until there are none left
+        Ncomplete = 0
+        while(Ncomplete < self.njobs):
+            # iterate over list of currently pending group of processes
+            for i in range(len(waitlist)):
+                # check for completion of all processes in this worker group
+                if self._checkreqlist(waitlist[i]):
+                    # all ranks group doing job corresponding to place i in waitlist
+                    # have returned; identify this worker group and remove it from the
+                    # waitlist and worker list
+                    worker = worker_groups[i]
+                    Ncomplete += 1
+                    waitlist.pop(i)
+                    worker_groups.pop(i)
+                    if nextjob < self.njobs:
+                        # more jobs to do; assign processes in group worker
+                        # the job with index nextjob, increment
+                        reqs=self._assign_job(worker,nextjob)
+                        waitlist.append(reqs)
+                        worker_groups.append(worker)
+                        nextjob += 1
+                    # waitlist has been modified so exit waitlist loop with break
+                    break
+
+        # no more jobs to assign; dismiss all processes in all groups by 
+        # assigning job = -1, causing all workers processes to return
+        for worker in range(self.Ngroups): 
+            self._assign_job(worker,-1)
+
+        return 
+    
+    def _work(self):
+        """
+        Listen for job assignments and run workfunc         
+        """
+        # listen for job assignments from the scheduler
+        while True:
+            self.comm.Recv(self.job_buff,source=0) # receive assignment from rank=0 scheduler
+            job = self.job_buff[0]                 # unpack job index
+            if job < 0: return                     # job < 0 means no more jobs to do
+            try:
+                self._workfunc(self.groupcomm,job) # call work function for job
+            except Exception as e:
+                self.log.error(f'FAILED: call to workfunc')
+                self.log.error(e)
```

Comment:
Commentary: I see that you also have a try/except within the spexec workfunc, but it is good for this general class to not trust the user to manage their own exceptions and be robust to that, even if it is somewhat redundant for this case.